Bluetooth ISO reconfigure CIG

Thalley opened the pull request:
Adds support to reconfigured a CIG after creation.

Thalley committed:
Bluetooth: ISO: Change CIG/BIG to use slist instead of array

A CIG may be updated later to include more CIS, and thus a slist
makes more sense.

The BIG doesn't need the change, but it makes more sense to
have similar handling for both.

This change also removes the requirement that the arrays
used to create the CIG/BIG need to static.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: ISO: Rename cis/bis in cig/big to cis/bis channels

Rename the field in the struct to x_channels to make
it more clear what the field represents.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: ISO: Move cis/bis param checks to earlier

Move the verification of the individual cis/bis checks to
an earlier point, because we actually start allocating
resources for the groups.

This removes a double check for NULL, while also
allows us to terminate earlier if there were any
issues.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: ISO: Rename bt_iso_cig_create_param

Renames the struct from bt_iso_cig_create_param to
bt_iso_cig_param as the same struct can, without
modification, be used to update the CIG as well
(function to support that will come in a later commit).

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: ISO: Move validation of CIG parameters to function

Move the validation of specific CIG parameters to its
own function. This is mainly to be able to reuse it for
validationg CIG refconfiguration parameters.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: ISO: Add CIG reconfigure function

Add function to reconfigure and even add additional
CIS to a CIG.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: ISO: Add CIG state machine

Add a state to the bt_iso_cig struct. This makes
checks for the CIG state easier, and follows the
state machine in the core spec.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo reviewed (changes requested):
Some minor comments, and possibly one issue to fix.
Also, the wording in the first commit message seems a bit unclear:
"This change also removes the requirement that the arrays used to create the CIG/BIG don't need to static."
I think the intention is to state that it removes the requirement that the arrays need to be static.

asbjornsabo reviewed (approved):
My comments have been resolved

cvinayak commented on include/bluetooth/iso.h:
```diff
@@ -515,9 +511,31 @@ int bt_iso_server_register(struct bt_iso_server *server);
+ *  in the call to bt_iso_cig_create(), these will be added to the group.
+ *  It is not possible to remove any CIS from the group after creation.
+ *
+ *  This can called at any time before connecting an ISO to a remote device.
```
```suggestion
 *  This can be called at any time before connecting an ISO to a remote device.
```

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -440,6 +442,28 @@ static void bt_iso_chan_disconnected(struct bt_iso_chan *chan, uint8_t reason)
+			}
+		}
+
+		if (fully_disconnected) {
```
Since the `if` clause checks `BT_ISO_CONNECTED` or `BT_ISO_CONNECT`, use `is_chan_connected` as the variable. This will ensure the boolean is assigned false, and `!is_chan_connected` be used here.

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1048,14 +1072,24 @@ static struct net_buf *hci_le_set_cig_params(const struct bt_iso_cig *cig,
+		return NULL;
+	}
+
+	return &cigs[iso_chan->iso->iso.cig_id];
```
Should there be a array out-of-bounds check too?

Thalley replied:
Arguably yes. I'll add an assert since this should not happen unless something else have gone wrong.

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1065,63 +1099,117 @@ static struct bt_iso_cig *get_free_cig(void)
+static int cig_init_cis(struct bt_iso_cig *cig,
+			const struct bt_iso_cig_param *param)
+{
+	for (int i = 0; i < param->num_cis; i++) {
```
Any reason to use `int` while other places `uint8_t` ?

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1065,63 +1099,117 @@ static struct bt_iso_cig *get_free_cig(void)
-		cis->iso->iso.cis_id = i;
+			cis->iso->iso.cig_id = cig->id;
+			cis->iso->iso.is_bis = false;
+			cis->iso->iso.cis_id = cig->num_cis++;
```
Dont you need a check for `cig->num_cis++` , I believe `CHECKIF` is used in the caller API (can that be compiled out?) , is it checked there to ensure there is no overflow here?

Thalley replied:
I am not fully sure what you would check here, that num_cis is larger than `CONFIG_BT_ISO_MAX_CHAN`?

cvinayak replied:
Yes, as `cig_init_cis` is in the user call path with return value, `cig->num_cis` should be check with `CONFIG_BT_ISO_MAX_CHAN`  and `-ENOMEM` be returned if stream ids are exhausted. I could not find where `cig->num_cis` is initialised or reset, it is designed to roll-over, should it be renamed as `cis_id_next` or `cis_id_curr`  if the purpose is  to assign unique CIS id?

Thalley replied:
If too many CIS are supplied, the above check will fail:
```
cis->iso = iso_new();
if (cis->iso == NULL) {
	BT_ERR("Unable to allocate CIS connection");
	return -ENOMEM;
}
```

In case of error, `cleanup_cig` would be called, but I see a problem with that, as we may also call `cleanup_cig` in the reconfigure function, which we shouldn't. I'll add an update for that part, but otherwise the remaining values should already have been checked.

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1452,59 +1590,38 @@ static struct bt_iso_big *big_lookup_flag(int bit)
-				return -EINVAL;
-			}
-		}
+	for (int i = 0; i < num_bis; i++) {
```
`uint8_t i` ?

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1595,8 +1719,26 @@ int bt_iso_big_create(struct bt_le_ext_adv *padv, struct bt_iso_big_create_param
 	}
 
 	for (int i = 0; i < param->num_bis; i++) {
```
uint8_t ?

cvinayak commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1966,8 +2104,25 @@ int bt_iso_big_sync(struct bt_le_per_adv_sync *sync, struct bt_iso_big_sync_para
 	}
 
 	for (int i = 0; i < param->num_bis; i++) {
```
uint8_t ?

Thalley committed:
Bluetooth: ISO: Update loop variables types in iso.c

Some of the loop variables used `int` where it
makes more sense to use `size_t` or `uint8_t`.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
@cvinayak I've updated the loop variable types in a new commit.

cfriedt commented:
@jhedberg - look OK?

jhedberg reviewed (approved):
> @jhedberg - look OK?

Yeah, from a code perspective this looks alright. It's also nicely split up into multiple commits and the commit messages seem descriptive enough. However, I don't have a way to runtime test this nor am I particularly familiar with the related Bluetooth specs.
